DOCUMENTATION.md

```diff
@@ -43,6 +45,190 @@ The overall territory is laid out here in this [C4](https://c4model.com) System
+- providing a clean approach to the testing of Reaction logic with and without involving your actual Event Store (the `MemoryStoreProjector` component is a key part of that story).
+- getting observability on messages handling and processing. Propulsion exposes diagnostic and telemetry information that enables effective monitoring. That's essential for tuning your Handlers for the production-ready system.
+
+## Overview of running projections with Propulsion
+
+### Glossary
+
+#### Tranches
+
+A Propulsion source can optionally split it's reading into multiple independent _Tranches_ (a tranche is a portion/division of a pool, it literally means slice in french). Each tranche is read (including reading up to a specified number of batches) and checkpointed independently. The internal Submitter component takes a batch from each Tranche in turn to ensure fair distribution of work across all Tranches that have work to be processed.
+
+### Supported Inputs
+
+Propulsion provides for processing events from the following sources:
+
+1. `Propulsion.CosmosStore`: Lightly wraps the Change Feed Processor library within the `Microsoft.Azure.Cosmos` client. It provides a load balanced leasing system (like Kafka), enabling multiple instances of a processor to parallelise processing. It allows to parallelise processing for multiple consumers. It's done by using [competing consumers pattern](https://www.enterpriseintegrationpatterns.com/patterns/messaging/CompetingConsumers.html) with the same Processor Name (Consumer Group Name). There can be as many active processors as physical partitions. The client API indicates the Partition Identity via the Context's `LeaseToken`).
+   - Each physical Partition maps to a Tranche, so processing still honors stream concurrency limits, even if the CFP library assigns more than one partition to a given instance.
+   - The CFP library provides for estimation of the Lag of documents to be processed between the current checkpoint position and the tail of each partition, which can surfaced as a metric.
+   - Over time, as a Container splits into more physical partitions (either due to assigning > 10K RU per physical partition, or by exceeding the 50GB/partition limit), more Partitions will arise.
+   - Checkpoints are maintained by the CFP library within an auxiliary (by convention, a sister container with the name`{container}-aux`) CosmosDB container (the document containing the checkpoint doubles as a lease management target)
+2. `Propulsion.DynamoStore`: Reads from an Index Table maintained by a `Propulsion.DynamoStore.Indexer` Lambda
+    - In the current implementation, the Indexer feeds all items into a single Tranche within its index (extending it to support multiple Tranches based on a hash of the stream identifier would not be difficult).
```

Comment:
**QUESTION:** What's _its index_?

Reply:
in DDB, there is a DynamoDB Streams log with 24h retention that sees all inserts, updates and deletes
DynamoStore provides a consumer for this which runs in Lambda. Per batch of <= 10000 (normally configured for <=1000) updates, it writes an entry into a separate Dynamo Table that functions as an Index for the actual Table. When you want to run reactions, you read the Index Table. You only need to go back to the original table if you need the event bodies
(This is all because DDB, Like Mongo has no way to walk all the data -- as opposed to cosmos, which is designed to enable this)

Reply:
Added some more details. Please let me know if its still incomplete (I dont want to provide too much duplication here as the point of this section is more to explain stream based dispatch to handlers and how to think about reactors)